File: sample-core/src/main/java/com/quickblox/sample/core/utils/imagepick/fragment/ImagePickHelperFragment.java
package com.quickblox.sample.core.utils.imagepick.fragment;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AppCompatActivity;

import com.quickblox.sample.core.utils.ImageUtils;
import com.quickblox.sample.core.utils.imagepick.GetFilepathFromUriTask;
import com.quickblox.sample.core.utils.imagepick.OnImagePickedListener;

public class ImagePickHelperFragment extends Fragment {

    private static final String ARG_REQUEST_CODE = "requestCode";
    private static final String ARG_PARENT_FRAGMENT = "parentFragment";

    private static final String TAG = ImagePickHelperFragment.class.getSimpleName();

    private OnImagePickedListener listener;

    public static ImagePickHelperFragment start(Fragment fragment, int requestCode) {
        Bundle args = new Bundle();
        args.putInt(ARG_REQUEST_CODE, requestCode);
        args.putString(ARG_PARENT_FRAGMENT, fragment.getClass().getSimpleName());

        return start(fragment.getActivity().getSupportFragmentManager(), args);
    }

    public static ImagePickHelperFragment start(FragmentActivity activity, int requestCode) {
        Bundle args = new Bundle();
        args.putInt(ARG_REQUEST_CODE, requestCode);

        return start(activity.getSupportFragmentManager(), args);
    }

    private static ImagePickHelperFragment start(FragmentManager fm, Bundle args) {
        ImagePickHelperFragment fragment = (ImagePickHelperFragment) fm.findFragmentByTag(TAG);
        if (fragment == null) {
            fragment = new ImagePickHelperFragment();
            fm.beginTransaction().add(fragment, TAG).commitAllowingStateLoss();
            fragment.setArguments(args);
        }
        return fragment;
    }

    public static void stop(FragmentManager fm) {
        Fragment fragment = fm.findFragmentByTag(TAG);
        if (fragment != null) {
            fm.beginTransaction().remove(fragment).commitAllowingStateLoss();
        }
    }

    public ImagePickHelperFragment() {}

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        Fragment fragment = ((AppCompatActivity) context).getSupportFragmentManager()
                .findFragmentByTag(getArguments().getString(ARG_PARENT_FRAGMENT));
        if (fragment != null) {
            if (fragment instanceof OnImagePickedListener) {
                listener = (OnImagePickedListener) fragment;
            }
        } else {
            if (context instanceof OnImagePickedListener) {
                listener = (OnImagePickedListener) context;
            }
        }

        if (listener == null) {
            throw new IllegalStateException("Either activity or fragment should implement OnImagePickedListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (isResultFromImagePick(requestCode, resultCode, data)) {
            if (requestCode == ImageUtils.CAMERA_REQUEST_CODE && (data == null || data.getData() == null)) {
                // Hacky way to get EXTRA_OUTPUT param to work.
                // When setting EXTRA_OUTPUT param in the camera intent there is a chance that data will return as null
                // So we just pass temporary camera file as a data, because RESULT_OK means that photo was written in the file.
                data = new Intent();
                data.setData(Uri.fromFile(ImageUtils.getLastUsedCameraFile()));
            }
            new GetFilepathFromUriTask(getChildFragmentManager(), listener, getArguments().getInt(ARG_REQUEST_CODE)).execute(data);
        } else {
            stop(getChildFragmentManager());
            if (listener != null) {
                listener.onImagePickClosed(getArguments().getInt(ARG_REQUEST_CODE));
            }
        }
    }

    public void setListener(OnImagePickedListener listener) {
        this.listener = listener;
    }

    private boolean isResultFromImagePick(int requestCode, int resultCode, Intent data) {
        return resultCode == Activity.RESULT_OK
                && ((requestCode == ImageUtils.CAMERA_REQUEST_CODE)
                || (requestCode == ImageUtils.GALLERY_REQUEST_CODE && data != null));
    }
}

File: sample-chat-kotlin/app/src/main/java/com/quickblox/sample/chat/kotlin/utils/imagepick/fragment/ImagePickHelperFragment.kt
package com.quickblox.sample.chat.kotlin.utils.imagepick.fragment

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import com.quickblox.sample.chat.kotlin.utils.CAMERA_REQUEST_CODE
import com.quickblox.sample.chat.kotlin.utils.GALLERY_REQUEST_CODE
import com.quickblox.sample.chat.kotlin.utils.getLastUsedCameraFile
import com.quickblox.sample.chat.kotlin.utils.imagepick.GetFilepathFromUriTask
import com.quickblox.sample.chat.kotlin.utils.imagepick.OnImagePickedListener

private const val ARG_REQUEST_CODE = "requestCode"
private const val ARG_PARENT_FRAGMENT = "parentFragment"

class ImagePickHelperFragment : Fragment() {

    private var listener: OnImagePickedListener? = null

    companion object {
        private val TAG = ImagePickHelperFragment::class.java.simpleName

        fun getInstance(activity: FragmentActivity, requestCode: Int): ImagePickHelperFragment {
            val fragmentManager = activity.supportFragmentManager
            var fragment: ImagePickHelperFragment? = fragmentManager.findFragmentByTag(TAG) as ImagePickHelperFragment?

            if (fragment == null) {
                fragment = ImagePickHelperFragment()
                fragmentManager.beginTransaction().add(fragment, TAG).commitAllowingStateLoss()
                val args = Bundle()
                args.putInt(ARG_REQUEST_CODE, requestCode)
                fragment.arguments = args
            }
            return fragment
        }

        fun stop(fragmentManager: FragmentManager) {
            val fragment = fragmentManager.findFragmentByTag(TAG)
            fragment?.let {
                fragmentManager.beginTransaction().remove(fragment).commitAllowingStateLoss()
            }
        }
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        val tag = arguments?.getString(ARG_PARENT_FRAGMENT)
        val fragment = (context as AppCompatActivity).supportFragmentManager.findFragmentByTag(tag)

        if (fragment != null && fragment is OnImagePickedListener) {
            listener = fragment
        } else if (context is OnImagePickedListener) {
            listener = context
        }
        if (listener == null) {
            throw IllegalStateException("Either activity or fragment should implement OnImagePickedListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        var intent = data
        super.onActivityResult(requestCode, resultCode, intent)
        if (isResultFromImagePick(requestCode, resultCode, intent)) {
            if (requestCode == CAMERA_REQUEST_CODE && (intent == null || intent.data == null)) {
                // Hacky way to get EXTRA_OUTPUT param to work.
                // When setting EXTRA_OUTPUT param in the camera intent there is a chance that data will return as null
                // So we just pass temporary camera file as a data, because RESULT_OK means that photo was written in the file.
                intent = Intent()
                intent.data = Uri.fromFile(getLastUsedCameraFile())
            }
            GetFilepathFromUriTask(childFragmentManager, listener, arguments!!.getInt(ARG_REQUEST_CODE)).execute(intent)
        } else {
            stop(childFragmentManager)
            listener?.onImagePickClosed(arguments!!.getInt(ARG_REQUEST_CODE))
        }
    }

    private fun isResultFromImagePick(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        return resultCode == Activity.RESULT_OK && (requestCode == CAMERA_REQUEST_CODE
                || requestCode == GALLERY_REQUEST_CODE && data != null)
    }
}